contracts/campaignAirdrops/ExistingERC20DropCampaign.sol

```diff
@@ -0,0 +1,126 @@
+    event Claimed(address indexed claimer);
+    event AirdropFunded(address contractAddress);
+
+    error AlreadyFunded();
+    error Unauthorized();
+    error AlreadyRedeemed();
+    error AirdropStillActive();
+    error AirdropExpired();
+    error NotEligible();
+    error InsufficientAmount();
+    error InsufficientLiquidity();
+    error MerkleRootAlreadySet();
+
+    modifier onlyAdmin() {
+        if (msg.sender != IAirBroFactory(airbroCampaignFactoryAddress).admin()) revert Unauthorized();
+        _;
+    }
+
+    constructor(
+        address _rewardToken,
+        uint256 _tokenSupply,
+        address _airbroCampaignFactoryAddress
+    ) {
+        rewardToken = IERC20(_rewardToken);
+        tokenSupply = _tokenSupply;
+        airbroCampaignFactoryAddress = _airbroCampaignFactoryAddress;
+    }
+
+    /// @notice Sets the merkleRoot and the number of claimers (also setting the amount each claimer receivers)
+    /// @notice can only be done by admin
+    /// @param _merkleRoot The root hash of the Merle Tree
+    /// @param _numberOfClaimers The number of people eligible to claim
+    function setMerkleRoot(bytes32 _merkleRoot, uint256 _numberOfClaimers) external onlyAdmin {
+        if (merkleRootSet) revert MerkleRootAlreadySet();
+
+        merkleRootSet = true;
+        merkleRoot = _merkleRoot;
+        airdropExpirationTimestamp = block.timestamp + 60 days;
+        numberOfClaimers = _numberOfClaimers;
+        tokensPerClaim = tokenSupply / numberOfClaimers;
+
+        emit MerkleRootChanged(_merkleRoot, _numberOfClaimers);
+    }
+
+    /// @notice Allows the airdrop creator to provide funds for the airdrop reward
+    function fundAirdrop() external {
+        if (airdropFunded) revert AlreadyFunded();
+
+        if (rewardToken.balanceOf(msg.sender) < tokenSupply) revert InsufficientAmount();
+        if (rewardToken.allowance(msg.sender, address(this)) < tokenSupply) revert InsufficientAmount();
+
+        airdropFunded = true;
+        airdropFunder = msg.sender;
+
+        rewardToken.safeTransferFrom(msg.sender, address(this), tokenSupply);
+        emit AirdropFunded(address(this));
+    }
+
+    /// @notice Allows the airdrop creator to withdraw back the funds after the airdrop has finished
+    function withdrawAirdropFunds() external {
+        if (airdropFunder != msg.sender) revert Unauthorized();
+        if (block.timestamp <= airdropExpirationTimestamp) revert AirdropStillActive();
+
+        rewardToken.safeTransfer(msg.sender, rewardToken.balanceOf(address(this)));
+    }
+
+    /// @notice Allows eligible users to claim their ERC20 airdrop
+    /// @param _merkleProof is the merkle proof that this user is eligible for claiming the ERC20 airdrop
+    function claim(bytes32[] calldata _merkleProof) external {
+        if (isEligibleForReward(_merkleProof)) {
+            if (rewardToken.balanceOf(address(this)) < tokensPerClaim) revert InsufficientLiquidity();
```

Comment:
safeTransfer will fail on 104 if you remove this line

Reply:
I'm not sure I understand the feedback / our mistake, could you please elaborate on what the mistake is?

Reply:
If you try `rewardToken.safeTransfer(msg.sender, tokenAmount);`
And `msg.sender`  doesn't have enough `tokenAmount` this call `rewardToken.safeTransfer(msg.sender, tokenAmount);` will revert.
So why double check something that is already checked in `safeTransfer`

Backend / Subgraph can monitor smart contract, and return data to frontend if there is not enough liquidity.
Here we're doing:
Balance checking before transfer
Transfer
    transfer internally is checking balance and reverting if fails

Reply:
Ok, so:
safeTransfer already does the checks, variables set will revert if **safeTransfer** reverts. With revert, no state is changed